### DISCUSSION TITLE
Grid with labels

### DISCUSSION THREAD

**Discussion by vdcast** [KoalaPlot/koalaplot-core#38] [Q&A]
Is it possible to group labels in this way?
It's like a grid with places for labels around chart.

![donut_chart_example](https://github.com/KoalaPlot/koalaplot-core/assets/108469609/fb531de0-0084-43ae-84a0-daea759eaeb7)

**Comment by vdcast** [comment]
For now I have this result. I checked documentation, but didn't find out how to setup the places for labels.


![image](https://github.com/KoalaPlot/koalaplot-core/assets/108469609/09571a21-cde5-46f1-ad56-7d6fef843218)

**Comment by gsteckman** [comment]
The current algorithm for placing the labels doesn't use a grid, but more like around a circle with a radius larger than the pie, and adjusted to avoid labels from overlapping. It might be possible to make the algorithm pluggable so others could implement different methods of laying out the labels. It would need investigation.

**Reply by vdcast** [reply]
Got it, but not fully :D 
Can I adjust current algorithm or plug my own? Or not yet?

**Reply by vdcast** [reply]
Also, some of labels are going outside and taking too much place in composition. Is there any way to adjust  label's place?

![image](https://github.com/KoalaPlot/koalaplot-core/assets/108469609/cfbb4887-cd1b-4dad-9895-30a87922a022)

**Reply by vdcast** [reply]
This time it went to the top

![image](https://github.com/KoalaPlot/koalaplot-core/assets/108469609/c6443714-bd2f-4b93-84b9-f5340d375b22)

**Comment by gsteckman** [comment]
It's not possible at this time to adjust the algorithm beyond the ```labelSpacing``` parameter or to plug in your own layout algorithm without writing a new version of the pie chart component.

It's not straightforward to have a general-purpose algorithm for all possible slice sizes and positions and label geometries. The current algorithm places the labels nominally on a circle with a diameter that is ```labelSpacing``` larger than the pie and then shifts them if they would overlap with labels previously placed (going clockwise). So in your last example, the slices were too close together to fit all the labels without overlapping, so it started pushing them up until they didn't overlap.

I have seen other pie chart label layout approaches that put all the labels in 2 columns, one on the left side and one on the right side, of the chart area. This is similar to your grid idea but without top/bottom rows. That would probably work reasonably well for the geometry of the labels you are using (I had considered it initially, but preferred the look of the circular layout). In the 2-column approach, it would place labels for slices with a center angle between [0,180) degrees on the right and between [180,360) degrees on the left (where 0 is at the top and angles increase clockwise), then space them vertically so they don't overlap and the connector lines don't cross over the pie. In your example for December, all those labels on the left would stay on the left but be in a column instead of a modified arc. There would still be more labels on the left than on the right because that is how the slices are being drawn.

So even with that approach, depending on the data and ordering of the slices you might end up with a tall column of labels on the left and as little as 1 label on the right. Another approach you could try is to order the slices according to their size, largest to smallest, and remove slices (data points) after a certain quantity threshold, say 10 slices, and have a final slice that is "other" that aggregates the values of what you removed. Or, remove slices below a specific value threshold, say 5% or 2%, and aggregate those into an "other" category.

Have you tried using a large value of ```labelSpacing```, like 2?

**Reply by vdcast** [reply]
I got your idea about 2 columns, but it's still not what I'm looking for.
This chart shows January records, so please ignore December text in the top of screenshot.

In 2nd screenshot I put in labelSpacing = 2f, you can see the result.

<details><summary>1f:</summary>
<p>

![image](https://github.com/KoalaPlot/koalaplot-core/assets/108469609/09b250bf-c388-4fe7-a20e-b0713e1d627a)

</p>
</details> 

<details><summary>2f:</summary>
<p>

![image](https://github.com/KoalaPlot/koalaplot-core/assets/108469609/40ad0bb8-668d-4332-b38d-bab70f96e6b2)

</p>
</details>

**Reply by vdcast** [reply]
What about vertical grid (4x5) with places:

 [1]   [2]  [3] [4]
[14]  [0]  [0] [5]
[13]  [0]  [0] [6]
[12]  [0]  [0] [7]
[11] [10] [9] [8]

Where:
[1]..[14] -  places for labels for chart's data, places are not const and can be changed due to chart's data for labels.
[0] - places for pie chart, places are const.

Is it possible with the lib?

**Reply by vdcast** [reply]
> Another approach you could try is to order the slices according to their size, largest to smallest, and remove slices (data points) after a certain quantity threshold, say 10 slices, and have a final slice that is "other" that aggregates the values of what you removed. Or, remove slices below a specific value threshold, say 5% or 2%, and aggregate those into an "other" category.

It's good idea and I will try to do it!
But I think it's still not guarantee the chart won't change. I mean chart's size and labels places.

**Reply by gsteckman** [reply]
I just committed a new version of ```PieChart``` that will take a ```LabelOffsetProvider``` as a parameter. This will allow anyone to implement custom label placement logic, provide it to the ```PieChart```, and then it will use the positions provided to place the labels.

If your use case is very specific and you know you will always have 14 labels and you want them in specific places, then the ```LabelOffsetProvider``` will allow you to specify where to put those labels relative to the center of the pie.

Regarding your grid approach, that might work as long as there are exactly 14 labels/slices. But what happens if there are only 10 labels, for example? Also, if the slice for label 13 is on the right side of the pie, for example, and its center point is at the 5 O'Clock position, then the connector line for it will either need to be wrapped around the outer circumference of the pie, or it will cross over the pie.
